BUG: fix to_latex() when using MultiIndex with NaN in (#14249)

tomneep opened the pull request:
- [x] closes #14249
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This attempt to close #14249 which actually reports two related issues (one in the issue one in the SO link) with to_latex() with MultiIndex when the MultiIndex contains a NaN. The output is now more consistent with what you'd get from to_csv() or to_html().

tomneep committed:
BUG: fix to_latex() when using MultiIndex with NaN in (#14249)

jreback commented:
@tomneep are you testing both issues in the issue?

jreback commented:
cc @toobaz if any comments

tomneep commented:
@jreback yes, I only added one test but parametrized it to cover both of the issues

toobaz commented:
> cc @toobaz if any comments

Sorry for the delay. I think the patch works as intended, but I also think that replacing ```lev2 = lev.format()``` and all the new ``blank`` definition with

```python
lev2 = self.frame.index.get_level_values(i).format()
```

would be cleaner.

toobaz commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -864,6 +864,7 @@ I/O
 - Bug in :func:`read_csv` where missing values were not being handled properly when ``keep_default_na=False`` with dictionary ``na_values`` (:issue:`19227`)
 - Bug in :func:`read_sas` where a file with 0 variables gave an ``AttributeError`` incorrectly. Now it gives an ``EmptyDataError`` (:issue:`18184`)
 - Bug in :func:`DataFrame.to_latex()` where pairs of braces meant to serve as invisible placeholders were escaped (:issue:`18667`)
+- Bug in :func:`DataFrame.to_latex()` where a ``NaN`` in a ``MultiIndex`` would cause an ``IndexError`` or incorrect output (:issue:`14249`)
```
where an all-``NaN`` level in a ``MultiIndex``

(I think)

tomneep replied:
In the SO link in #14249 there is a case where `NaN`s are being silently replaced with other non-`NaN` values in the level and this fixes both that issue and the one actually reported

toobaz replied:
right!

tomneep commented:
Hi @toobaz, 

Thanks for checking the patch, I wanted to keep the changes as minimal as possible as it is my first contribution. 
I agree that this isn't so clean so I had a go at refactoring that section of code- see
d63e88fa92617a6ba9121297e8a3a8447d3dfe7e. 

The tests all pass so if you prefer this approach (I certainly do) I can make a PR for that commit instead (maybe after a bit of tidying).

tomneep commented:
Hi @jreback @toobaz 
Since I see that https://github.com/pandas-dev/pandas/pull/20032 is also addressing the same area of code to this, I'd like to get some clarification to see if we can move this forward. 

The rewrite I mention in the comment above closes the issue this PR initially aimed to close as well as #19981, and an (as far as I can tell unreported) issue that `to_latex(sparsify=False)` is still sparsifying regardless in the master. My impression is that this would be a good place for #20032 to start making changes but I would like you opinion. 

The main thing I don't know, is whether I should create a new PR since the changes are now more substantial than this initial one, or if I can just keep using this one and commit the changes to this branch?

jreback commented:
can you rebase. we moved some files around.

tomneep commented:
Ok, rebased. I also rebased my alternative branch that tidies this portion of the code up: c0daa15d06ba9d2c63b34a1148cc65d433c6027a

My preference would be to merge that branch instead (in which case a couple more tests could be added) but let me know which you prefer.

toobaz commented:
@tomneep again sorry for the delay. I also think the alternative branch is much better: please feel free to open another PR.

(Please also clarify whether that branch is incompatible with/alternative to #20032, or just requires some merge work)

tomneep commented:
I've opened #20797 for the improved branch. 

Pehaps it will mean some of #20032 won't be needed but I don't think this is completely incompatible.
